File: benchmarks/cross-language/java/PingPong.java
// Java ping-pong benchmark using threads and BlockingQueue
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicLong;

public class PingPong {
    private static int getMessages() {
        String env = System.getenv("BENCHMARK_MESSAGES");
        return env != null ? Integer.parseInt(env) : 100000;
    }
    private static final int MESSAGES = getMessages();

    static class PingThread extends Thread {
        private final BlockingQueue<Integer> sendQueue;
        private final BlockingQueue<Integer> recvQueue;

        public PingThread(BlockingQueue<Integer> send, BlockingQueue<Integer> recv) {
            this.sendQueue = send;
            this.recvQueue = recv;
        }

        @Override
        public void run() {
            try {
                for (int i = 0; i < MESSAGES; i++) {
                    sendQueue.put(i);
                    int received = recvQueue.take();
                    // VALIDATE: Must receive echo of what we sent
                    if (received != i) {
                        System.err.println("ERROR: Ping sent " + i + " but got back " + received);
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    static class PongThread extends Thread {
        private final BlockingQueue<Integer> sendQueue;
        private final BlockingQueue<Integer> recvQueue;

        public PongThread(BlockingQueue<Integer> send, BlockingQueue<Integer> recv) {
            this.sendQueue = send;
            this.recvQueue = recv;
        }

        @Override
        public void run() {
            try {
                for (int i = 0; i < MESSAGES; i++) {
                    int received = recvQueue.take();
                    // VALIDATE: Must receive expected sequence
                    if (received != i) {
                        System.err.println("ERROR: Pong expected " + i + " but got " + received);
                    }
                    // Echo back what we received
                    sendQueue.put(received);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== Java Ping-Pong Benchmark ===");
        System.out.println("Messages: " + MESSAGES);
        System.out.println("Using Java threads with ArrayBlockingQueue");
        System.out.println();

        BlockingQueue<Integer> queueA = new ArrayBlockingQueue<>(1);
        BlockingQueue<Integer> queueB = new ArrayBlockingQueue<>(1);

        PingThread ping = new PingThread(queueA, queueB);
        PongThread pong = new PongThread(queueB, queueA);

        long start = System.nanoTime();

        ping.start();
        pong.start();

        ping.join();
        pong.join();

        long end = System.nanoTime();
        long totalNs = end - start;

        double nsPerMsg = (double) totalNs / MESSAGES;
        double throughput = 1e9 / nsPerMsg;

        System.out.printf("ns/msg:         %.2f\n", nsPerMsg);
        System.out.printf("Throughput:     %.2f M msg/sec\n", throughput / 1e6);
    }
}

File: benchmarks/cross-language/rust/src/ping_pong.rs
use std::sync::mpsc::sync_channel;
use std::thread;
use std::time::Instant;

fn get_messages() -> usize {
    std::env::var("BENCHMARK_MESSAGES")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(100_000)
}

fn main() {
    let messages = get_messages();
    println!("=== Rust Ping-Pong Benchmark ===");
    println!("Messages: {}", messages);
    println!("Using std::thread with sync_channel\n");

    let (ping_tx, ping_rx) = sync_channel::<i32>(1);
    let (pong_tx, pong_rx) = sync_channel::<i32>(1);

    let start = Instant::now();

    let ping_thread = thread::spawn(move || {
        for i in 0..messages {
            let val = i as i32;
            ping_tx.send(val).unwrap();
            let received = pong_rx.recv().unwrap();
            if received != val {
                eprintln!("ERROR: Ping sent {} but got back {}", val, received);
            }
        }
    });

    let pong_thread = thread::spawn(move || {
        for i in 0..messages {
            let expected = i as i32;
            let received = ping_rx.recv().unwrap();
            if received != expected {
                eprintln!("ERROR: Pong expected {} but got {}", expected, received);
            }
            pong_tx.send(received).unwrap();
        }
    });

    ping_thread.join().unwrap();
    pong_thread.join().unwrap();

    let elapsed = start.elapsed().as_secs_f64();
    let ns_per_msg = elapsed * 1e9 / messages as f64;
    let msg_per_sec = messages as f64 / elapsed;

    println!("ns/msg:         {:.2}", ns_per_msg);
    println!("Throughput:     {:.2} M msg/sec", msg_per_sec / 1_000_000.0);
}
